src/basic/xattr-util.c

```diff
@@ -156,9 +156,18 @@ int getxattr_at_malloc(
 
                 r = getxattr_pinned_internal(fd, path, at_flags, by_procfs, name, v, l);
                 if (r >= 0) {
+                        /* Refuse extended attributes with embedded NUL bytes if the caller isn't interested
+                         * in the size. After all this must mean the caller assumes we return a NUL
+                         * terminated strings, but if there's a NUL byte embedded they are definitely not
+                         * regular strings */
+                        if (!ret_size && r > 1 && memchr(v, 0, r - 1))
+                                return -EBADMSG;
+
                         v[r] = 0; /* NUL terminate */
                         *ret = TAKE_PTR(v);
-                        return r;
+                        if (ret_size)
+                                *ret_size = r;
```

Comment:
It's no longer necessary for getxattr_pinned_internal() to carry the INT_MAX constraint then? Just inline the function again and use the ssize_t we got directly?